pkg/contract/solana/gateway_message.go

```diff
@@ -0,0 +1,73 @@
+package solana
+
+import (
+	"encoding/binary"
+
+	"github.com/ethereum/go-ethereum/crypto"
+	"github.com/gagliardetto/solana-go"
+)
+
+// MsgWithdraw is the message for the Solana gateway withdraw instruction
+type MsgWithdraw struct {
+	// ChainID is the chain ID of Solana chain
+	ChainID uint64
+
+	// Nonce is the nonce for the withdraw
+	Nonce uint64
+
+	// Amount is the lamports amount for the withdraw
+	Amount uint64
+
+	// To is the recipient address for the withdraw
+	To solana.PublicKey
+
+	// Signature is the signature of the message
+	Signature [65]byte
+}
+
+// NewMsgWithdraw returns a new MsgWithdraw
+func NewMsgWithdraw(chainID, nonce, amount uint64, to solana.PublicKey) *MsgWithdraw {
+	return &MsgWithdraw{
+		ChainID: chainID,
+		Nonce:   nonce,
+		Amount:  amount,
+		To:      to,
+	}
+}
+
+// Hash packs the withdraw message and computes the hash
+func (msg *MsgWithdraw) Hash() [32]byte {
+	var message []byte
+	buff := make([]byte, 8)
+
+	binary.BigEndian.PutUint64(buff, msg.ChainID)
+	message = append(message, buff...)
+
+	binary.BigEndian.PutUint64(buff, msg.Nonce)
+	message = append(message, buff...)
+
+	binary.BigEndian.PutUint64(buff, msg.Amount)
+	message = append(message, buff...)
+
+	message = append(message, msg.To.Bytes()...)
+
+	return crypto.Keccak256Hash(message)
+}
+
+// WithSignature attaches the signature to the message
+func (msg *MsgWithdraw) WithSignature(signature [65]byte) *MsgWithdraw {
+	msg.Signature = signature
```

Comment:
- Should we make `msg.signature` unexported and add `msg.Signature()` method? Otherwise someone else can alter it by a direct access instead of this method.
- Consider naming it `SetSignature(...)`

Reply:
Agree on the renaming. `With...` should be used for options. Here this is not an option, a function called in the normal workflow